python/cudf/cudf/core/dataframe.py

```diff
@@ -349,18 +355,45 @@ def _setitem_tuple_arg(self, key, value):
                         )
                     else:
                         raise ValueError(
-                            f"shape mismatch: value array of shape "
-                            f"{value_df.shape} could not be "
-                            f"broadcast to indexing result of shape "
-                            f"{columns_df.shape}"
+                            template.substitute(
+                                value1=value_df.shape,
+                                value2=columns_df.shape,
+                            )
                         )
                 else:
                     value_cols = value_df._data.columns
                 for i, col in enumerate(columns_df._column_names):
                     self._frame[col].loc[key[0]] = value_cols[i]
             else:
-                for col in columns_df._column_names:
-                    self._frame[col].loc[key[0]] = value
+                if isinstance(value, cudf.DataFrame):
+                    if value.shape != self._frame.loc[key[0]].shape:
+                        raise ValueError(
+                            template.substitute(
+                                value1=value.shape,
+                                value2=self._frame.loc[key[0]].shape,
+                            )
+                        )
+                    value_column_names = set(value._column_names)
+                    scatter_map = _indices_from_labels(self._frame, key[0])
+                    for col in columns_df._column_names:
+                        columns_df[col][scatter_map] = (
+                            value._data[col]
+                            if col in value_column_names
+                            else cudf.NA
+                        )
+                else:
+                    value = np.array(value)
+                    value = value.reshape((-1, value.shape[0]))
+                    if value.shape != self._frame.loc[key[0]].shape:
+                        raise ValueError(
+                            template.substitute(
+                                value1=value.shape,
+                                value2=self._frame.loc[key[0]].shape,
+                            )
+                        )
+                    scatter_map = _indices_from_labels(self._frame, key[0])
+                    for i, col in enumerate(columns_df._column_names):
+                        self._frame._data[col][scatter_map] = value[:, i]
```

Comment:
Just a thought, an attempt to reduce the duplication here, can you try below?
```suggestion
                scatter_map = _indices_from_labels(self._frame, key[0])
                self._frame.iloc[scatter_map] = value
```